File: golang/src/reflexion_agent.go
package main

import (
	"fmt"
	"log"
)

// ReflexionAgent implements the Reflexion pattern for codebase analysis
type ReflexionAgent struct {
	*BaseAgent
	memoryRestorationNeeded bool
	originalSystemMessage   *Message
}

// NewReflexionAgent creates a new Reflexion agent
func NewReflexionAgent(modelName string, baseURL string) *ReflexionAgent {
	return &ReflexionAgent{
		BaseAgent:               NewBaseAgent(modelName, baseURL, ReflexionSystemPrompt),
		memoryRestorationNeeded: false,
		originalSystemMessage:   nil,
	}
}

// Run executes the Reflexion agent's analysis process
func (a *ReflexionAgent) Run(prompt string, directory string) (string, error) {
	if err := a.InitializeMemory(prompt, directory); err != nil {
		return "", fmt.Errorf("error initializing memory: %v", err)
	}

	maxSteps := 15
	for step := 1; step <= maxSteps; step++ {
		log.Printf("\n--- Step %d ---\n", step)
		
		// Check if we need to restore the system message
		if a.memoryRestorationNeeded && a.originalSystemMessage != nil {
			// Restore original system message
			for i := range a.Memory {
				if a.Memory[i].Role == "system" {
					a.Memory[i] = *a.originalSystemMessage
					break
				}
			}
			a.memoryRestorationNeeded = false
		}
		
		// Call the LLM
		_, err := a.CallLLM()
		if err != nil {
			return "", fmt.Errorf("error in step %d: %v", step, err)
		}
		
		// Check if we have a final answer
		if a.FinalAnswer != "" {
			log.Println("Final answer received")
			break
		}
		
		// Check if we have tool calls
		if len(a.Memory[len(a.Memory)-1].ToolCalls) == 0 {
			log.Println("No tool calls received")
			continue
		}
		
		// Execute each tool call
		for _, toolCall := range a.Memory[len(a.Memory)-1].ToolCalls {
			log.Printf("Executing tool: %s\n", toolCall.Function.Name)
			
			_, err := a.ExecuteTool(toolCall)
			if err != nil {
				return "", fmt.Errorf("error executing tool %s: %v", toolCall.Function.Name, err)
			}
			
			// Check if we have a final answer after executing the tool
			if a.FinalAnswer != "" && toolCall.Function.Name == "final_answer" {
				log.Println("Final answer received from tool execution")
				break
			}
		}
		
		// Check if we have a final answer after this step
		if a.FinalAnswer != "" {
			break
		}
		
		// Add reflection step after each iteration (except the last one)
		if step < maxSteps && a.FinalAnswer == "" {
			if err := a.AddReflection(); err != nil {
				return "", fmt.Errorf("error adding reflection: %v", err)
			}
			
			// Schedule restoration for next iteration
			if step+1 < maxSteps {
				a.memoryRestorationNeeded = true
			}
		}
		
		log.Printf("Memory length: %d messages\n", len(a.Memory))
	}
	
	// Save the final answer
	if a.FinalAnswer == "" {
		return "", fmt.Errorf("no final answer generated after maximum steps")
	}
	
	outputFile, err := a.SaveFinalAnswer()
	if err != nil {
		return "", fmt.Errorf("error saving final answer: %v", err)
	}
	
	log.Printf("Analysis complete. Results saved to %s\n", outputFile)
	
	return outputFile, nil
}

// AddReflection adds a reflection step to the agent's memory
func (a *ReflexionAgent) AddReflection() error {
	// Find the system message in memory and update it
	for i := range a.Memory {
		if a.Memory[i].Role == "system" {
			// Store original system message if not already stored
			if a.originalSystemMessage == nil {
				originalMsg := a.Memory[i]
				a.originalSystemMessage = &originalMsg
			}
			
			// Add reflection instruction to system message
			reflectionInstruction := "\n\nBefore responding, reflect on your previous actions. Were they effective? How can you improve your approach? Incorporate these reflections into your response."
			a.Memory[i].Content += reflectionInstruction
			break
		}
	}
	
	log.Println("Added reflection to memory")
	
	return nil
}

File: rust/src/agents/reflexion_agent.rs
use anyhow::{Context, Result};
use log::{debug, error, info, warn};
use serde_json::Value;
use std::path::Path;

use crate::agents::base_agent::{Agent, BaseAgent};
use crate::prompts;
use crate::types::{Message, Role, ToolCall};

/// Agent that uses the Reflexion pattern for codebase analysis
pub struct ReflexionAgent {
    /// The base agent implementation
    pub base: BaseAgent,
    
    /// The original system prompt before reflection
    original_system_prompt: String,
    
    /// Flag indicating if the system prompt needs to be restored
    system_prompt_modified: bool,
}

impl ReflexionAgent {
    /// Create a new Reflexion agent
    pub fn new(model_name: &str, base_url: Option<&str>) -> Result<Self> {
        let system_prompt = prompts::reflexion_system_prompt();
        
        Ok(Self {
            base: BaseAgent::new(model_name, base_url, &system_prompt)?,
            original_system_prompt: system_prompt,
            system_prompt_modified: false,
        })
    }
    
    /// Add a reflection step after tool execution
    fn add_reflection(&mut self) -> Result<()> {
        // Store the original system prompt if this is the first reflection
        if !self.system_prompt_modified {
            self.original_system_prompt = self.base.system_prompt.clone();
        }
        
        // Temporarily modify the system prompt to include reflection instructions
        let reflection_prompt = format!(
            "{}\n\n{}",
            self.original_system_prompt,
            prompts::REFLECTION_PROMPT
        );
        
        // Update the system prompt
        self.base.system_prompt = reflection_prompt;
        self.system_prompt_modified = true;
        
        // Call the LLM to generate a reflection
        let (content, _) = self.call_llm()?;
        
        // Add the reflection to memory
        self.base.memory.push(Message::assistant(&content));
        
        info!("Added reflection to memory");
        info!("Memory length: {} messages", self.base.memory.len());
        
        // Restore the original system prompt for the next iteration
        self.base.system_prompt = self.original_system_prompt.clone();
        self.system_prompt_modified = false;
        
        Ok(())
    }
}

impl Agent for ReflexionAgent {
    fn run(&mut self, prompt: &str, directory: &str) -> Result<String> {
        info!("Using Reflexion agent with reflection capabilities");
        
        // Initialize memory with the prompt and directory
        self.initialize_memory(prompt, directory);
        
        let mut step = 1;
        
        // Main agent loop
        loop {
            info!("\n--- Step {} ---", step);
            step += 1;
            
            // Call the LLM
            let (content, tool_calls) = self.call_llm()?;
            
            // Add the assistant's message to memory
            let mut assistant_message = Message::assistant(&content);
            assistant_message.tool_calls = tool_calls.clone();
            self.base.memory.push(assistant_message);
            
            // If no tool calls, continue to the next iteration
            if tool_calls.is_empty() {
                info!("No tool calls received");
                continue;
            }
            
            // Process each tool call
            for tool_call in &tool_calls {
                let tool_name = &tool_call.function.name;
                info!("Executing tool: {}", tool_name);
                
                // Execute the tool
                let result = self.execute_tool(tool_call)?;
                
                // If the tool is final_answer, we're done
                if tool_name == "final_answer" {
                    info!("Final answer received from tool execution");
                    self.base.final_answer = Some(result.clone());
                    return Ok(result);
                }
                
                // Add the tool result to memory
                self.base.memory.push(Message::tool(
                    &tool_call.id,
                    tool_name,
                    &result,
                ));
            }
            
            // Add a reflection step after tool execution
            self.add_reflection()?;
            
            // Check if we've exceeded the maximum number of steps
            if step > 20 {
                warn!("Exceeded maximum number of steps (20)");
                return Err(anyhow::anyhow!("Exceeded maximum number of steps"));
            }
        }
    }
    
    fn initialize_memory(&mut self, prompt: &str, directory: &str) {
        self.base.initialize_memory(prompt, directory);
    }
    
    fn call_llm(&self) -> Result<(String, Vec<ToolCall>)> {
        self.base.call_llm()
    }
    
    fn execute_tool(&self, tool_call: &ToolCall) -> Result<String> {
        self.base.execute_tool(tool_call)
    }
}
